src/baseline.rs

```diff
@@ -4,69 +4,95 @@ use anyhow::Context as _;
 use crates_index::Crate;
 
 use crate::dump::RustDocCommand;
+use crate::manifest::Manifest;
 use crate::util::slugify;
 use crate::GlobalConfig;
 
-mod crate_features {
-    /// Sometimes crates ship types with fields or variants that are included
-    /// only when certain features are enabled.
-    ///
-    /// By default, we want to generate rustdoc with `--all-features`,
-    /// but that option isn't available outside of the current crate,
-    /// so we have to implement it ourselves.
-
-    pub(crate) fn get_all_crate_features_from_registry(
-        crate_: &crates_index::Version,
-    ) -> Vec<String> {
-        // Implicit features from optional dependencies have to be added separately
-        // from regular features: https://github.com/obi1kenobi/cargo-semver-checks/issues/265
-        let mut implicit_features: std::collections::BTreeSet<_> = crate_
-            .dependencies()
-            .iter()
-            .filter_map(|dep| dep.is_optional().then_some(dep.name()))
-            .map(|x| x.to_string())
-            .collect();
-        for feature_defn in crate_.features().values().flatten() {
-            // "If you specify the optional dependency with the dep: prefix anywhere
-            //  in the [features] table, that disables the implicit feature."
-            // https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies
-            if let Some(optional_dep) = feature_defn.strip_prefix("dep:") {
-                implicit_features.remove(optional_dep);
-            }
-        }
-        let regular_features: std::collections::BTreeSet<_> =
-            crate_.features().keys().cloned().collect();
-        let mut all_crate_features = implicit_features;
-        all_crate_features.extend(regular_features);
-        all_crate_features.into_iter().collect()
-    }
-
-    #[allow(unused_variables)]
-    pub(crate) fn get_all_crate_features_from_manifest(path: &std::path::Path) -> Vec<String> {
-        unimplemented!()
-    }
-}
-
-#[allow(dead_code)]
 #[derive(Debug, Clone)]
 enum CrateSource<'a> {
     Registry { crate_: &'a crates_index::Version },
-    ManifestPath { path: &'a Path, name: String },
+    ManifestPath { manifest: &'a Manifest },
 }
 
 impl<'a> CrateSource<'a> {
-    fn name(&self) -> &str {
+    fn name(&self) -> anyhow::Result<String> {
+        Ok(match self {
+            Self::Registry { crate_ } => crate_.name().to_string(),
+            Self::ManifestPath { manifest } => crate::manifest::get_package_name(manifest)?,
+        })
+    }
+
+    fn version(&self) -> anyhow::Result<String> {
+        Ok(match self {
+            Self::Registry { crate_ } => crate_.version().to_string(),
+            Self::ManifestPath { manifest } => crate::manifest::get_package_version(manifest)?,
+        })
+    }
```

Comment:
I'm tempted to say we should go in the other direction on these methods: make the `crate::manifest::get_*` functions return `Result<&'str>` instead of making this function return `Result<String>`. What do you think?

Reply:
Done in efd3b5884a5e4872781379b6e392dce3d3db5dc0.